include/glow/Partitioner/Partitioner.h

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_PARTITIONER_PARTITIONER_H
+#define GLOW_PARTITIONER_PARTITIONER_H
+
+#include "glow/Graph/Context.h"
+#include "glow/Graph/Graph.h"
+
+#include <map>
+#include <set>
+#include <string>
+
+namespace glow {
+
+using ModuleList = std::list<Module *>;
+
+struct DependencyDAG {
+  /// List of modules after partitioning. Each module will be assigned to run a
+  /// device by provioner.
+  ModuleList modules_;
+  /// The dependencies of each module.
+  Map modulesDAG_;
+  /// The root of this DAG.
+  Modules *root;
+  /// The number of nodes in the DAG.
+  int k;
+};
+
+/// Here we assume all the devices are identical.
+struct DeviceInfo {
+  /// Number of the devices.
+  int num;
+  /// Memory constrains of each device. unit GB.
+  int memSize;
+  /// TO DO... communication constrins and so on
+}
+
+/// Partition
+class Partitioner {
+
+  /// Dependency map type.
+  using Map = llvm::DenseMap<Module *, ModuleList>;
+
+  /// The module need to be decomposed.
+  Module *module_;
+
+  /// The cost model related to device.
+  DeviceInfo *deviceInfo_;
+
+  /// The result of module partitioning.
+  DependencyDAG dependencyDAG_;
+
+public:
+  /// \p parent is the module which contains the functions need to be divided.
+  /// Here we assume that all the functions in one module belong to a same
+  /// "Function Family", that is, without considerting the "dynamic stuff" (i.e.
+  /// batch size, input/output shape of each op), all the functions are
+  /// identical. The required memory and computation cost for each op can be
+  /// found in Module. The \p devices provides the cost model related to
+  /// devices.
+  Partitioner(const Module &parent, const DeviceInfo &devices);
```

Comment:
Do we want to partition all of the functions in the module or just one function?

Reply:
We will partition all functions. However, according to current discussion, a module may contains a list of functions, but these functions belong to a same "function family". That is, those functions reflect same ML model, but may with different batch or tensor shape (e.g. dynamic shape in caffe2).